textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/Join.java

```diff
@@ -42,14 +88,160 @@ public void open() throws Exception {
 	 */
 	@Override
 	public ITuple getNextTuple() throws Exception {
-		// TODO Auto-generated method stub
+		if(innerTupleList.isEmpty()) {
+			return null;
+		}
+
 		ITuple nextTuple = null;
+
+		do {
+			if(getOuterOperatorNextTuple == true) {
+				if((outerTuple = outerOperator.getNextTuple()) != null) {
+					getOuterOperatorNextTuple = false;
+				} else {
+					return null;
+				}
+			}
+
+			if (innerOperatorCursor <= innerTupleList.size() - 1) {
+				innerTuple = innerTupleList.get(innerOperatorCursor);
+				innerOperatorCursor++;
+				if(innerOperatorCursor == innerTupleList.size()) {
+					innerOperatorCursor = 0;
+					getOuterOperatorNextTuple = true;
+				}
+			}
+
+			nextTuple = processTuples(outerTuple, innerTuple, joinPredicate);
+		} while(nextTuple == null);
+
 		return nextTuple;
 	}
 
 	@Override
 	public void close() throws Exception {
-		// TODO Auto-generated method stub
-		
+		try {
+			outerOperator.close();
+			// innerOperator.close(); already called in open()
+
+		} catch(Exception e) {
+			e.printStackTrace();
+			throw new DataFlowException(e.getMessage(), e);
+		}
+		// Clear the inner tuple list from memory on close.
+		innerTupleList.clear();
+	}
+
+	// Used to compare IDs of the tuples.
+	private boolean compareId(ITuple outerTuple, ITuple innerTuple) {
+		// TODO(Flavio): what if the join predicate has an invalid id attribute?
+		// First check if the field in question exits by using try catch.
+		// Will throw an exception if it doesn't exist. This leads to return false.
+		// Then check if both the fields are of type IntegerField.
+		// --> This is the bare minimum thing that can be done to to verify valid 
+		// id attribute (as of now) (probably it is better to add a field called ID).
+		String fieldName = joinPredicate.getidAttribute().getFieldName();
```

Comment:
We are forcing user to add an ID attribute for every Tuple. I think this is pretty reasonable solution. But we need to discuss how should we enforce this constraint.